pkg/ext/stores/useractivity/store.go

```diff
@@ -0,0 +1,312 @@
+		return nil, err
+	}
+
+	// crafting UserActivity from requested Token name.
+	ua := &ext.UserActivity{
+		ObjectMeta: metav1.ObjectMeta{
+			CreationTimestamp: activityToken.CreationTimestamp,
+			Name:              name,
+		},
+		Status: ext.UserActivityStatus{},
+	}
+
+	if activityToken.ActivityLastSeenAt != nil {
+		ua.Status.ExpiresAt = activityToken.ActivityLastSeenAt.String()
+	} else {
+		ua.Status.ExpiresAt = metav1.Time{}.String()
+	}
+
+	return ua, nil
+}
+
+// userFrom is a helper that extracts and validates the user info from the request's context.
+func (s *Store) userFrom(ctx context.Context) (k8suser.Info, error) {
+	userInfo, ok := request.UserFrom(ctx)
+	if !ok {
+		return nil, apierrors.NewInternalError(fmt.Errorf("missing user info"))
+	}
+
+	userName := userInfo.GetName()
+
+	if strings.Contains(userName, ":") { // E.g. system:admin
+		return nil, apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("user %s is not a Rancher user", userName))
+	}
+
+	if !slices.Contains(userInfo.GetGroups(), GroupCattleAuthenticated) {
+		return nil, apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("user %s is not a Rancher user", userName))
+	}
+
+	user, err := s.userCache.Get(userName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("user %s not found", userName))
+		}
+
+		return nil, apierrors.NewInternalError(fmt.Errorf("error getting user %s: %w", userName, err))
+	}
+
+	if user.Enabled != nil && !*user.Enabled {
+		return nil, apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("user %s is disabled", userName))
+	}
+
+	return userInfo, nil
+}
+
+func first(values []string) string {
+	if len(values) > 0 {
+		return values[0]
+	}
+	return ""
+}
+
+func validateActivityToken(auth, activity *v3Legacy.Token) error {
+	// verify auth and activity token have the same userID
+	if auth.UserID != activity.UserID {
+		return apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("request token %s and activity token %s have different users", auth.Name, activity.Name))
+	}
+
+	// verify auth and activity token has the same auth provider
+	if auth.AuthProvider != activity.AuthProvider {
+		return apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("request token %s and activity token %s have different auth providers", auth.Name, activity.Name))
+	}
+
+	// verify auth and activity token has the same auth user principal
+	if auth.UserPrincipal.Name != activity.UserPrincipal.Name {
+		return apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("request token %s and activity token %s have different user principals", auth.Name, activity.Name))
+	}
+
+	// verify activity token is of the expected kind
+	if activity.Labels[TokenKind] != "session" {
+		return apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("activity token %s is not a session token", activity.Name))
+	}
+	return nil
```

Comment:
Are we missing a check for the activity token being disabled? I though we had this check  in a previous iteration.
```suggestion

	if activity.Enabled != nil && !*activity.Enabled {
		return apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("activity token %s is disabled", activity.Name))
	}

	return nil
```

Reply:
we have, but just for user:
```
	if user.Enabled != nil && !*user.Enabled {
		return nil, apierrors.NewForbidden(GVR.GroupResource(), "", fmt.Errorf("user %s is disabled", userName))
	}
```